File: graph/bellman_ford.c
#include <stdio.h>
#include <limits.h>

// Edge 정의
typedef struct {
    int src;
    int dst;
    int weight;
} Edge;

// Vertex 정의
typedef struct {
    int dist;
    int pi;
} Vertex;

// 단일 소스 초기화
void initializeSingleSource(Vertex vertices[], int size, int source) {
    for (int i = 0; i < size; i++) {
        vertices[i].dist = INT_MAX;
        vertices[i].pi = -1;
    }
    vertices[source].dist = 0;
}

// Relaxation 함수
void relax(Vertex vertices[], const Edge* edge) {
    if (vertices[edge->src].dist != INT_MAX && vertices[edge->dst].dist > vertices[edge->src].dist + edge->weight) {
        vertices[edge->dst].dist = vertices[edge->src].dist + edge->weight;
        vertices[edge->dst].pi = edge->src;
    }
}


// 음수 가중치 사이클 감지 함수
int detectNegativeCycle(const Vertex vertices[], const Edge edges[], int numEdges) {
    for (int i = 0; i < numEdges; i++) {
        if (vertices[edges[i].src].dist != INT_MAX && vertices[edges[i].dst].dist > vertices[edges[i].src].dist + edges[i].weight) {
            return 1; // Negative weight cycle detected
        }
    }
    return 0; // No negative weight cycle
}


// Bellman-Ford 알고리즘
int bellmanFord(Vertex vertices[], const Edge edges[], int numVertices, int numEdges, int source) {
    initializeSingleSource(vertices, numVertices, source);

    for (int i = 1; i < numVertices; i++) {
        for (int j = 0; j < numEdges; j++) {
            relax(vertices, &edges[j]);
        }
    }

    for (int i = 0; i < numEdges; i++) {
        if (vertices[edges[i].src].dist != INT_MAX && vertices[edges[i].dst].dist > vertices[edges[i].src].dist + edges[i].weight) {
            return 0; // Negative weight cycle detected
        }
    }

    return 1; // Success
}

int main() {
    Vertex vertices[5];
    Edge edges[] = {
        {0, 1, 6}, {0, 3, 7},
        {1, 2, 5}, {1, 3, 8}, {1, 4, -4},
        {2, 1, -2},
        {3, 2, -3}, {3, 4, 9},
        {4, 0, 2}, {4, 2, 7}
    };

    int numVertices = 5;
    int numEdges = sizeof(edges) / sizeof(edges[0]);
    int source = 0;

    if (bellmanFord(vertices, edges, numVertices, numEdges, source)) {
        printf("Shortest path from vertex %d:\n", source);
        for (int i = 0; i < numVertices; i++) {
            printf("Vertex %d: Distance=%d, Predecessor=%d\n", i, vertices[i].dist, vertices[i].pi);
        }
    } else {
        printf("Negative weight cycle detected. Bellman-Ford algorithm failed.\n");
    }

    return 0;
}

File: graph/bellman_ford.py
import sys

class Edge:
    src = 0
    dst = 0
    weight = 0

    def __init__(self, src: int, dst: int, weight: int):
        self.src = src
        self.dst = dst
        self.weight = weight


class Vertex:
    dist = 0
    pi = 0

    def __init__(self):
        self.dist = sys.maxsize
        self.pi = -1


def initializeSingleSource(vertices, source: int):
    for vertex in vertices:
        vertex.dist = sys.maxsize
        vertex.pi = -1
    vertices[source].dist = 0

def relax(vertices, edge):
    if vertices[edge.src].dist != sys.maxsize and vertices[edge.dst].dist > vertices[edge.src].dist + edge.weight:
        vertices[edge.dst].dist = vertices[edge.src].dist + edge.weight
        vertices[edge.dst].pi = edge.src

def bellmanFord(vertices, edges, source):
    initializeSingleSource(vertices, source)

    for i in range(1, len(vertices)+1):
        for edge in edges:
            relax(vertices, edge)
    
    for edge in edges:
        if vertices[edge.src].dist != sys.maxsize and vertices[edge.dst].dist > vertices[edge.src].dist + edge.weight:
            return False
    
    return True


def detectNegativeCycle(vertices, edges):
    for edge in edges:
        if vertices[edge.src].dist != sys.maxsize and vertices[edge.dst].dist > vertices[edge.src].dist + edge.weight:
            return True  # Negative weight cycle detected
    return False  # No negative weight cycle


def main():
    vertices = [Vertex() for i in range(5)]
    edges = [
        Edge(0,1,6), Edge(0,3,7),
        Edge(1,2,5), Edge(1,3,8), Edge(1,4,-4),
        Edge(2,1,-2),
        Edge(3,2,-3), Edge(3,4,9),
        Edge(4,0,2), Edge(4,2,7)
    ]

    source = 0

    if bellmanFord(vertices, edges, source):
        print("Shortest path from vertex", source, ':')
        for i in range(len(vertices)):
            print("Vertex", i, ": Distance =", vertices[i].dist, ", Predecessor =", vertices[i].pi)

    else:
        print("Negative weight cycle detected. Bellman-Ford algorithm failed.")

    detectNegativeCycle(vertices,edges)
if __name__ == '__main__':
    main()